config/dynamic.c

```diff
@@ -16,11 +16,535 @@
+			"dyn_cfg: Failed to write data to the socket connection - (%s)\n",
+			strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+static int
+reply_client_message(int conn_fd, const char *reply_msg, uint16_t reply_len)
+{
+	int ret;
+	uint16_t nlen = htons(reply_len);
+
+	/* The first two bytes: the length of the message in network order. */
+	ret = write_nbytes(conn_fd, (char *)&nlen, sizeof(nlen));
+	if (ret != 0)
+		return -1;
+
+	/* Sending the message. */
+	ret = write_nbytes(conn_fd, reply_msg, reply_len);
+	if (ret != 0)
+		return -1;
+
+	return 0;
+}
+
+static int
+process_client_message(int conn_fd,
+       const char *msg, int msg_len, lua_State *lua_state)
+{
+	/*
+	 * TODO Implement the functionalities to process the clients' request.
+	 *
+	 * Gatekeeper: Adding, listing, and removing BP components;
+	 *
+	 * Gatekeeper: Adding, listing, removing IP ranges
+	 * handled by the global LPM table in GKs;
+	 *
+	 * Gatekeeper and Grantor: Listing the ARP and ND table.
+	 * This is important for network diagnosis.
+	 *
+	 * Grantor: Updating the policy GTs run.
+	 */
+
+	int ret;
+	size_t reply_len;
+	const char *reply_msg;
+
+	if (msg_len == 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: the received message is an empty string\n"); 
+		return -1;
+	}
+
+	/* Load the client's Lua chunk, and run it. */
+	ret = luaL_loadbuffer(lua_state, msg, msg_len, "message")
+		|| lua_pcall(lua_state, 0, 1, 0);
+	if (ret != 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: %s!\n", lua_tostring(lua_state, -1));
+		return -1;
+	}
+
+	reply_msg = luaL_checklstring(lua_state, -1, &reply_len);
+	if (reply_msg == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: dy_cmd_proc() returns a NULL string!\n");
+		reply_client_message(conn_fd,
+			CLIENT_PROC_ERROR, strlen(CLIENT_PROC_ERROR));
+		return -1;
+	}
+
+	if (reply_len > MSG_MAX_LEN) {
+		RTE_LOG(ERR, MALLOC,
+			"dyn_cfg: The reply message length (%lu) exceeds the limit!\n",
+			reply_len);
+		reply_len = MSG_MAX_LEN;
+	}
+
+	return reply_client_message(conn_fd, reply_msg, reply_len);
+}
+
+/*
+ * Return values:
+ * 	0: Read @nbytes successfully.
+ * 	-1: Connection closed by the client.
+ */
+static int
+read_nbytes(int conn_fd, char *msg_buff, int nbytes)
+{
+	int recv_size;
+	int tot_size = 0;
+
+	while ((recv_size = read(conn_fd, msg_buff + tot_size,
+			nbytes - tot_size)) > 0) {
+		tot_size += recv_size;
+		if (tot_size >= nbytes)
+			break;
+	}
+
+	/* The connection with the client is closed. */
+	if (recv_size == 0) {
+		RTE_LOG(WARNING, GATEKEEPER,
+			"dyn_cfg: Client disconnected\n");
+		return -1;
+	}
+
+	if (recv_size < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: Failed to read data from the socket connection - (%s)\n",
+			strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+/*
+ * Return values:
+ * 	-1: Error happens.
+ * 	0: Command successfully processed, may need to process further commands.
+ * 	1: Connection closed by the client or some read/write errors happened.
```

Comment:
I don't see the reason for having the return `1` since it is the same as returning `-1`.